allennlp/data/iterators/basic_iterator.py

```diff
@@ -22,27 +25,111 @@ class BasicIterator(DataIterator):
+        """
+        Take the next `max_instances` instances from the given dataset.
+        If `max_instances` is `None`, then just take all instances from the dataset.
+        If `max_instances` is not `None`, each call resumes where the previous one
+        left off, and when you get to the end of the dataset you start again from the beginning.
+        """
+        # If max_instances isn't specified, just iterate once over the whole dataset
+        if max_instances is None:
+            yield from iter(instances)
+        else:
+            # If we don't have a cursor for this dataset, create one. We use ``id()``
+            # for the key because ``instances`` could be a list, which can't be used as a key.
+            key = id(instances)
+            iterator = self._cursors.get(key, iter(instances))
+
+            while max_instances > 0:
+                try:
+                    # If there are instances left on this iterator,
+                    # yield one and decrement max_instances.
+                    yield next(iterator)
+                    max_instances -= 1
+                except StopIteration:
+                    # None left, so start over again at the beginning of the dataset.
+                    iterator = iter(instances)
+
+            # We may have a new iterator, so update the cursor.
+            self._cursors[key] = iterator
+
+    def _memory_sized_lists(self, instances: Iterable[Instance]) -> Iterable[List[Instance]]:
+        """
+        Breaks the dataset into "memory-sized" lists of instances,
+        which it yields up one at a time until it gets through a full epoch.
+
+        For example, if the dataset is already an in-memory list, and each epoch
+        represents one pass through the dataset, it just yields back the dataset.
+        Whereas if the dataset is lazily read from disk and we've specified to
+        load 1000 instances at a time, then it yields lists of 1000 instances each.
+        """
+        lazy = is_lazy(instances)
+
+        # Get an iterator over the next epoch worth of instances.
+        iterator = self._take_instances(instances, self._instances_per_epoch)
+
+        # We have four different cases to deal with:
+
+        # With lazy instances and no guidance about how many to load into memory,
+        # we just load ``batch_size`` instances at a time:
+        if lazy and self._max_instances_in_memory is None:
+            yield from lazy_groups_of(iterator, self._batch_size)
+        # If we specified max instances in memory, lazy or not, we just
+        # load ``max_instances_in_memory`` instances at a time:
+        elif self._max_instances_in_memory is not None:
+            yield from lazy_groups_of(iterator, self._max_instances_in_memory)
+        # If we have non-lazy instances, and we want all instances each epoch,
+        # then we just yield back the list of instances:
+        elif self._instances_per_epoch is None:
```

Comment:
Do you need this case?  It seems like `self._take_instances` will already do what you want if `self._instances_per_epoch` is None, and you can just collapse this with the following case.  Am I missing something?

Reply:
you're right that we could just return `list(iterator)` in both cases, but treating them separately avoids an unnecessary list copy (return instances as-is vs iterate and create a new list)